File: DSA/DSA_450/Cpp/14_DP/028_Max_Length_Chain_Pairs.cpp
/*
Problem: Maximum Length Chain of Pairs
URL: https://practice.geeksforgeeks.org/problems/max-length-chain/1

Problem Statement:
You are given N pairs of numbers. In every pair, the first number is always smaller than the second number. A pair (c, d) can follow another pair (a, b) if b < c. Chain of pairs can be formed in this fashion. Find the longest chain which can be formed from a given set of pairs.

Sample Input/Output:
Input: [[5,24], [39,60], [15,28], [27,40], [50,90]]
Output: 3
*/

#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    int Chain_DP(vector<pair<int, int>>& pairs, int n) {
        /*
        DP approach
        Time Complexity: O(n^2)
        Space Complexity: O(n)
        */
        sort(pairs.begin(), pairs.end());
        vector<int> dp(n, 1);
        for (int i = 1; i < n; i++) {
            for (int j = 0; j < i; j++) {
                if (pairs[j].second < pairs[i].first) {
                    dp[i] = max(dp[i], dp[j] + 1);
                }
            }
        }
        return *max_element(dp.begin(), dp.end());
    }

    int Chain_Greedy(vector<pair<int, int>>& pairs, int n) {
        /*
        Greedy approach
        Time Complexity: O(n log n)
        Space Complexity: O(1)
        */
        sort(pairs.begin(), pairs.end(), [](pair<int,int>& a, pair<int,int>& b) {
            return a.second < b.second;
        });
        int count = 1;
        int last = pairs[0].second;
        for (int i = 1; i < n; i++) {
            if (pairs[i].first > last) {
                count++;
                last = pairs[i].second;
            }
        }
        return count;
    }
};

void Test_Chain() {
    Solution solution;
    vector<pair<int, int>> pairs = {{5,24}, {39,60}, {15,28}, {27,40}, {50,90}};
    cout << "DP: " << solution.Chain_DP(pairs, pairs.size()) << endl;
    cout << "Greedy: " << solution.Chain_Greedy(pairs, pairs.size()) << endl;
}

int main() {
    Test_Chain();
    return 0;
}

File: DSA/DSA_450/Python/14_DP/028_Max_Length_Chain_Pairs.py
"""
Problem: Maximum Length Chain of Pairs
URL: https://practice.geeksforgeeks.org/problems/max-length-chain/1

Problem Statement:
You are given N pairs of numbers. In every pair, the first number is always smaller than the second number. A pair (c, d) can follow another pair (a, b) if b < c. Chain of pairs can be formed in this fashion. Find the longest chain which can be formed from a given set of pairs.

Sample Input/Output:
Input: [[5,24], [39,60], [15,28], [27,40], [50,90]]
Output: 3
"""

class Solution:
    def Chain_DP(self, pairs, n):
        """
        DP approach
        Time Complexity: O(n^2)
        Space Complexity: O(n)
        """
        pairs.sort()
        dp = [1] * n
        for i in range(1, n):
            for j in range(i):
                if pairs[j][1] < pairs[i][0]:
                    dp[i] = max(dp[i], dp[j] + 1)
        return max(dp)

    def Chain_Greedy(self, pairs, n):
        """
        Greedy approach
        Time Complexity: O(n log n)
        Space Complexity: O(1)
        """
        pairs.sort(key=lambda x: x[1])
        count = 1
        last = pairs[0][1]
        for i in range(1, n):
            if pairs[i][0] > last:
                count += 1
                last = pairs[i][1]
        return count

def Test_Chain():
    solution = Solution()
    pairs = [[5,24], [39,60], [15,28], [27,40], [50,90]]
    print("DP:", solution.Chain_DP(pairs, len(pairs)))
    print("Greedy:", solution.Chain_Greedy(pairs, len(pairs)))

if __name__ == "__main__":
    Test_Chain()
